File: example/jchatbot/src/main/java/zw/co/dcl/jchatbot/service/WebhookEventProcessor.java
package zw.co.dcl.jchatbot.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import zw.co.dcl.jchatbot.Util;
import zw.co.dcl.jchatbot.WebhookEvent;

@Service
public class WebhookEventProcessor {
    @Autowired
    private WebhookConfigService config;

    @Async
    @EventListener
    public void handleWebhook(WebhookEvent event) {
        config
                .getEntryInstance()
                .processWebhook(
                        event.getPayload(),
                        Util.requestHeadersToMap(event.getRequest())
                );
    }
}

File: example/kchatbot/src/main/kotlin/zw/co/dcl/jawce/chatbot/service/WebhookEventProcessor.kt
package zw.co.dcl.jawce.chatbot.service

import jakarta.servlet.http.HttpServletRequest
import org.springframework.context.ApplicationEvent
import org.springframework.context.event.EventListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import zw.co.dcl.jawce.chatbot.requestHeadersToMap

class WebhookEvent(
    source: Any,
    val payload: Map<String, Any>,
    val request: HttpServletRequest
) : ApplicationEvent(source) {}

@Service
class WebhookEventProcessor(private val config: WebhookConfigService) {
    @Async
    @EventListener
    fun handleWebhookEvent(webhookEvent: WebhookEvent) {
        config
            .entryInstance()
            .processWebhook(
                webhookEvent.payload,
                requestHeadersToMap(webhookEvent.request)
            )
    }
}
